File: backend/Services/GroupAccessService.cs
using Microsoft.EntityFrameworkCore;
using CrmArrighi.Data;
using CrmArrighi.Models;

namespace CrmArrighi.Services
{
    public interface IGroupAccessService
    {
        Task<bool> CanAccessModuleAsync(int usuarioId, string modulo);
        Task<bool> CanAccessScreenAsync(int usuarioId, string screenName);
        Task<List<string>> GetAccessibleModulesAsync(int usuarioId);
        Task<List<string>> GetAccessibleScreensAsync(int usuarioId);
        Task<bool> IsModuleHiddenAsync(int usuarioId, string modulo);
        Task<GroupAccessInfo> GetGroupAccessInfoAsync(int usuarioId);
    }

    public class GroupAccessInfo
    {
        public int GrupoId { get; set; }
        public string GrupoNome { get; set; } = string.Empty;
        public string Descricao { get; set; } = string.Empty;
        public List<string> ModulosPermitidos { get; set; } = new();
        public List<string> ModulosOcultos { get; set; } = new();
        public List<string> TelasPermitidas { get; set; } = new();
        public List<string> TelasOcultas { get; set; } = new();
        public bool ApenasFilial { get; set; }
        public bool ApenasLeitura { get; set; }
        public bool OcultarAbaUsuarios { get; set; }
    }

    public class GroupAccessService : IGroupAccessService
    {
        private readonly CrmArrighiContext _context;

        public GroupAccessService(CrmArrighiContext context)
        {
            _context = context;
        }

        public async Task<bool> CanAccessModuleAsync(int usuarioId, string modulo)
        {
            var groupInfo = await GetGroupAccessInfoAsync(usuarioId);
            return groupInfo.ModulosPermitidos.Contains(modulo);
        }

        public async Task<bool> CanAccessScreenAsync(int usuarioId, string screenName)
        {
            var groupInfo = await GetGroupAccessInfoAsync(usuarioId);
            return groupInfo.TelasPermitidas.Contains(screenName);
        }

        public async Task<List<string>> GetAccessibleModulesAsync(int usuarioId)
        {
            var groupInfo = await GetGroupAccessInfoAsync(usuarioId);
            return groupInfo.ModulosPermitidos;
        }

        public async Task<List<string>> GetAccessibleScreensAsync(int usuarioId)
        {
            var groupInfo = await GetGroupAccessInfoAsync(usuarioId);
            return groupInfo.TelasPermitidas;
        }

        public async Task<bool> IsModuleHiddenAsync(int usuarioId, string modulo)
        {
            var groupInfo = await GetGroupAccessInfoAsync(usuarioId);
            return groupInfo.ModulosOcultos.Contains(modulo);
        }

        public async Task<GroupAccessInfo> GetGroupAccessInfoAsync(int usuarioId)
        {
            var usuario = await _context.Usuarios
                .Include(u => u.GrupoAcesso)
                .FirstOrDefaultAsync(u => u.Id == usuarioId);

            if (usuario?.GrupoAcesso == null)
            {
                return GetUsuarioGroupInfo(); // Grupo Usuario (sem permissões)
            }

            return usuario.GrupoAcesso.Nome switch
            {
                "Usuario" => GetUsuarioGroupInfo(),
                "Administrador" => GetAdministradorGroupInfo(),
                "Consultores" => GetConsultoresGroupInfo(),
                "Administrativo de Filial" => GetAdministrativoFilialGroupInfo(),
                "Gestor de Filial" => GetGestorFilialGroupInfo(),
                "Cobrança e Financeiro" => GetCobrancaFinanceiroGroupInfo(),
                "Faturamento" => GetFaturamentoGroupInfo(),
                _ => GetUsuarioGroupInfo()
            };
        }

        private GroupAccessInfo GetUsuarioGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 1,
                GrupoNome = "Usuario",
                Descricao = "Usuário sem grupo de acesso",
                ModulosPermitidos = new List<string>(),
                ModulosOcultos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente", "Contrato", "Consultor", "Parceiro", "Boleto", "Usuario", "Filial", "GrupoAcesso", "Permissao" },
                TelasPermitidas = new List<string>(),
                TelasOcultas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes", "contratos", "consultores", "parceiros", "boletos", "usuarios", "filiais", "grupos-acesso", "permissoes" },
                ApenasFilial = false,
                ApenasLeitura = true,
                OcultarAbaUsuarios = true
            };
        }

        private GroupAccessInfo GetAdministradorGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 2,
                GrupoNome = "Administrador",
                Descricao = "Acesso total ao sistema",
                ModulosPermitidos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente", "Contrato", "Consultor", "Parceiro", "Boleto", "Usuario", "Filial", "GrupoAcesso", "Permissao" },
                ModulosOcultos = new List<string>(),
                TelasPermitidas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes", "contratos", "consultores", "parceiros", "boletos", "usuarios", "filiais", "grupos-acesso", "permissoes" },
                TelasOcultas = new List<string>(),
                ApenasFilial = false,
                ApenasLeitura = false,
                OcultarAbaUsuarios = false
            };
        }

        private GroupAccessInfo GetConsultoresGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 3,
                GrupoNome = "Consultores",
                Descricao = "Acesso a pessoa física/jurídica total, clientes da mesma filial e sem contrato",
                ModulosPermitidos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente" },
                ModulosOcultos = new List<string> { "Contrato", "Consultor", "Parceiro", "Boleto", "Usuario", "Filial", "GrupoAcesso", "Permissao" },
                TelasPermitidas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes" },
                TelasOcultas = new List<string> { "contratos", "consultores", "parceiros", "boletos", "usuarios", "filiais", "grupos-acesso", "permissoes" },
                ApenasFilial = true,
                ApenasLeitura = false,
                OcultarAbaUsuarios = true
            };
        }

        private GroupAccessInfo GetAdministrativoFilialGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 4,
                GrupoNome = "Administrativo de Filial",
                Descricao = "Apenas visualização de consultores, clientes e contratos da sua filial",
                ModulosPermitidos = new List<string> { "Consultor", "Cliente", "Contrato" },
                ModulosOcultos = new List<string> { "PessoaFisica", "PessoaJuridica", "Parceiro", "Boleto", "Usuario", "Filial", "GrupoAcesso", "Permissao" },
                TelasPermitidas = new List<string> { "consultores", "clientes", "contratos" },
                TelasOcultas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "parceiros", "boletos", "usuarios", "filiais", "grupos-acesso", "permissoes" },
                ApenasFilial = true,
                ApenasLeitura = true,
                OcultarAbaUsuarios = true
            };
        }

        private GroupAccessInfo GetGestorFilialGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 5,
                GrupoNome = "Gestor de Filial",
                Descricao = "Edita, inclui e exclui em todo o sistema porém somente na sua filial",
                ModulosPermitidos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente", "Contrato", "Consultor", "Parceiro", "Boleto", "Filial" },
                ModulosOcultos = new List<string> { "Usuario", "GrupoAcesso", "Permissao" },
                TelasPermitidas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes", "contratos", "consultores", "parceiros", "boletos", "filiais" },
                TelasOcultas = new List<string> { "usuarios", "grupos-acesso", "permissoes" },
                ApenasFilial = true,
                ApenasLeitura = false,
                OcultarAbaUsuarios = true
            };
        }

        private GroupAccessInfo GetCobrancaFinanceiroGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 6,
                GrupoNome = "Cobrança e Financeiro",
                Descricao = "Acesso total para visualizar todo o sistema (aba usuários oculta)",
                ModulosPermitidos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente", "Contrato", "Consultor", "Parceiro", "Boleto", "Filial", "GrupoAcesso", "Permissao" },
                ModulosOcultos = new List<string> { "Usuario" },
                TelasPermitidas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes", "contratos", "consultores", "parceiros", "boletos", "filiais", "grupos-acesso", "permissoes" },
                TelasOcultas = new List<string> { "usuarios" },
                ApenasFilial = false,
                ApenasLeitura = true,
                OcultarAbaUsuarios = true
            };
        }

        private GroupAccessInfo GetFaturamentoGroupInfo()
        {
            return new GroupAccessInfo
            {
                GrupoId = 7,
                GrupoNome = "Faturamento",
                Descricao = "Acesso similar ao administrador exceto módulo de usuários",
                ModulosPermitidos = new List<string> { "PessoaFisica", "PessoaJuridica", "Cliente", "Contrato", "Consultor", "Parceiro", "Boleto", "Filial", "GrupoAcesso", "Permissao" },
                ModulosOcultos = new List<string> { "Usuario" },
                TelasPermitidas = new List<string> { "pessoas-fisicas", "pessoas-juridicas", "clientes", "contratos", "consultores", "parceiros", "boletos", "filiais", "grupos-acesso", "permissoes" },
                TelasOcultas = new List<string> { "usuarios" },
                ApenasFilial = false,
                ApenasLeitura = false,
                OcultarAbaUsuarios = true
            };
        }
    }
}

File: frontend/src/services/groupAccess.service.ts
import { apiClient } from "@/lib/api";
import { GroupAccessInfo } from "@/types/groupAccess";

/**
 * Serviço para gerenciar grupos de acesso
 */
class GroupAccessService {
  private cache: {
    groupInfo?: GroupAccessInfo;
    lastFetch?: number;
    accessibleModules?: string[];
    accessibleScreens?: string[];
  } = {};

  private readonly CACHE_DURATION = 5 * 60 * 1000; // 5 minutos

  /**
   * Verifica se o cache está válido
   */
  private isCacheValid(): boolean {
    if (!this.cache.lastFetch) return false;
    return Date.now() - this.cache.lastFetch < this.CACHE_DURATION;
  }

  /**
   * Limpa o cache
   */
  private clearCache(): void {
    this.cache = {};
  }

  /**
   * Obtém informações completas sobre o acesso do usuário
   */
  async getGroupAccessInfo(): Promise<GroupAccessInfo> {
    if (this.isCacheValid() && this.cache.groupInfo) {
      return this.cache.groupInfo;
    }

    try {
      const response = await apiClient.get<GroupAccessInfo>(
        "/GroupAccess/user-info"
      );

      if (response.data) {
        this.cache.groupInfo = response.data;
        this.cache.lastFetch = Date.now();
        return response.data;
      }

      throw new Error("Dados de grupo não encontrados");
    } catch (error) {
      console.error("Erro ao obter informações do grupo:", error);
      // Retornar informações padrão para usuário sem grupo
      return {
        grupoId: 1,
        grupoNome: "Usuario",
        descricao: "Usuário sem grupo de acesso",
        modulosPermitidos: [],
        modulosOcultos: [],
        telasPermitidas: [],
        telasOcultas: [],
        apenasFilial: false,
        apenasLeitura: true,
        ocultarAbaUsuarios: true,
      };
    }
  }

  /**
   * Verifica se o usuário pode acessar um módulo específico
   */
  async canAccessModule(modulo: string): Promise<boolean> {
    try {
      const response = await apiClient.get<boolean>(
        `/GroupAccess/can-access-module/${modulo}`
      );
      return response.data ?? false;
    } catch (error) {
      console.error(`Erro ao verificar acesso ao módulo ${modulo}:`, error);
      return false;
    }
  }

  /**
   * Verifica se o usuário pode acessar uma tela específica
   */
  async canAccessScreen(screenName: string): Promise<boolean> {
    try {
      const response = await apiClient.get<boolean>(
        `/GroupAccess/can-access-screen/${screenName}`
      );
      return response.data ?? false;
    } catch (error) {
      console.error(`Erro ao verificar acesso à tela ${screenName}:`, error);
      return false;
    }
  }

  /**
   * Obtém lista de módulos que o usuário pode acessar
   */
  async getAccessibleModules(): Promise<string[]> {
    if (this.isCacheValid() && this.cache.accessibleModules) {
      return this.cache.accessibleModules;
    }

    try {
      const response = await apiClient.get<string[]>(
        "/GroupAccess/accessible-modules"
      );

      if (response.data) {
        this.cache.accessibleModules = response.data;
        this.cache.lastFetch = Date.now();
        return response.data;
      }

      return [];
    } catch (error) {
      console.error("Erro ao obter módulos acessíveis:", error);
      return [];
    }
  }

  /**
   * Obtém lista de telas que o usuário pode acessar
   */
  async getAccessibleScreens(): Promise<string[]> {
    if (this.isCacheValid() && this.cache.accessibleScreens) {
      return this.cache.accessibleScreens;
    }

    try {
      const response = await apiClient.get<string[]>(
        "/GroupAccess/accessible-screens"
      );

      if (response.data) {
        this.cache.accessibleScreens = response.data;
        this.cache.lastFetch = Date.now();
        return response.data;
      }

      return [];
    } catch (error) {
      console.error("Erro ao obter telas acessíveis:", error);
      return [];
    }
  }

  /**
   * Verifica se um módulo está oculto para o usuário
   */
  async isModuleHidden(modulo: string): Promise<boolean> {
    try {
      const response = await apiClient.get<boolean>(
        `/GroupAccess/is-module-hidden/${modulo}`
      );
      return response.data ?? false;
    } catch (error) {
      console.error(
        `Erro ao verificar se módulo ${modulo} está oculto:`,
        error
      );
      return true; // Por segurança, considerar como oculto em caso de erro
    }
  }

  /**
   * Verifica se uma tela está oculta para o usuário
   */
  async isScreenHidden(screenName: string): Promise<boolean> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.telasOcultas.includes(screenName);
  }

  /**
   * Verifica se o usuário tem acesso apenas à sua filial
   */
  async isFilialOnly(): Promise<boolean> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.apenasFilial;
  }

  /**
   * Verifica se o usuário tem apenas permissões de leitura
   */
  async isReadOnly(): Promise<boolean> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.apenasLeitura;
  }

  /**
   * Verifica se a aba de usuários deve ser ocultada
   */
  async shouldHideUsersTab(): Promise<boolean> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.ocultarAbaUsuarios;
  }

  /**
   * Obtém o nome do grupo do usuário
   */
  async getGroupName(): Promise<string> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.grupoNome;
  }

  /**
   * Obtém a descrição do grupo do usuário
   */
  async getGroupDescription(): Promise<string> {
    const groupInfo = await this.getGroupAccessInfo();
    return groupInfo.descricao;
  }

  /**
   * Invalida o cache (útil após mudanças de permissões)
   */
  invalidateCache(): void {
    this.clearCache();
  }

  /**
   * Verifica se o usuário é administrador
   */
  async isAdmin(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Administrador";
  }

  /**
   * Verifica se o usuário é consultor
   */
  async isConsultor(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Consultores";
  }

  /**
   * Verifica se o usuário é gestor de filial
   */
  async isGestorFilial(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Gestor de Filial";
  }

  /**
   * Verifica se o usuário é administrativo de filial
   */
  async isAdministrativoFilial(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Administrativo de Filial";
  }

  /**
   * Verifica se o usuário é do financeiro
   */
  async isFinanceiro(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Cobrança e Financeiro";
  }

  /**
   * Verifica se o usuário é do faturamento
   */
  async isFaturamento(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName === "Faturamento";
  }

  /**
   * Verifica se o usuário tem grupo válido (não é apenas "Usuario")
   */
  async hasValidGroup(): Promise<boolean> {
    const groupName = await this.getGroupName();
    return groupName !== "Usuario";
  }
}

// Instância singleton do serviço
export const groupAccessService = new GroupAccessService();
